// perfume.js/dist/lib/getNavigationTiming.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getNavigationTiming = void 0;
var constants_1 = require("./constants");
var isSupported_1 = require("./isSupported");
/**
 * Navigation Timing API provides performance metrics for HTML documents.
 * w3c.github.io/navigation-timing/
 * developers.google.com/web/fundamentals/performance/navigation-and-resource-timing
 */
var getNavigationTiming = function () {
    if (!(0, isSupported_1.isPerformanceSupported)()) {
        return {};
    }
    // There is an open issue to type correctly getEntriesByType
    // github.com/microsoft/TypeScript/issues/33866
    var n = constants_1.WP.getEntriesByType('navigation')[0];
    // In Safari version 11.2 Navigation Timing isn't supported yet
    if (!n) {
        return {};
    }
    var responseStart = n.responseStart;
    var responseEnd = n.responseEnd;
    // We cache the navigation time for future times
    return {
        // fetchStart marks when the browser starts to fetch a resource
        // responseEnd is when the last byte of the response arrives
        fetchTime: responseEnd - n.fetchStart,
        // Service worker time plus response time
        workerTime: n.workerStart > 0 ? responseEnd - n.workerStart : 0,
        // Request plus response time (network only)
        totalTime: responseEnd - n.requestStart,
        // Response time only (download)
        downloadTime: responseEnd - responseStart,
        // Time to First Byte (TTFB)
        timeToFirstByte: responseStart - n.requestStart,
        // HTTP header size
        headerSize: n.transferSize - n.encodedBodySize || 0,
        // Measuring DNS lookup time
        dnsLookupTime: n.domainLookupEnd - n.domainLookupStart,
        // redirects could add latency to requests
        redirectTime: n.redirectEnd - n.redirectStart,
    };
};
exports.getNavigationTiming = getNavigationTiming;
