// perfume.js/dist/lib/getNavigatorInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getNavigatorInfo = void 0;
var constants_1 = require("./constants");
var getNetworkInformation_1 = require("./getNetworkInformation");
var isLowEnd_1 = require("./isLowEnd");
/**
 * Information coming from window.navigator:
 * 1. Device Memory
 * 2. Hardware Concurency
 * 3. Status of the service worker:
 *    - controlled: a service worker is controlling the page
 *    - supported: the browser supports service worker
 *    - unsupported: the user's browser does not support service worker
 */
var getNavigatorInfo = function () {
    if (constants_1.WN) {
        return {
            deviceMemory: (0, constants_1.getDM)() || 0,
            hardwareConcurrency: (0, constants_1.getHC)() || 0,
            serviceWorkerStatus: 'serviceWorker' in constants_1.WN
                ? constants_1.WN.serviceWorker.controller
                    ? 'controlled'
                    : 'supported'
                : 'unsupported',
            isLowEndDevice: (0, isLowEnd_1.getIsLowEndDevice)(),
            isLowEndExperience: (0, isLowEnd_1.getIsLowEndExperience)(getNetworkInformation_1.et, getNetworkInformation_1.sd)
        };
    }
    return {};
};
exports.getNavigatorInfo = getNavigatorInfo;
